### PR TITLE
Add jdwp support

### PR DESCRIPTION
**Pull request by Daquiver1**
This PR tackles issue #56.
It creates a module for the JDWP commands and errors over [here](https://docs.oracle.com/javase/8/docs/platform/jpda/jdwp/jdwp-protocol.html#JDWP_Error)

### CODE REVIEW

File: jdwp/common.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+"""Basic types for JDWP messages."""
+
+from typing import List
+
+
+class Types:
```
Comment (michalgr) [inline jdwp/common.py:6]: Might be a good idea to turn this into an enum, perhaps `PrimitiveType`. Later we will need to include some more complex types, such as arrays and tagged unions and it might make sense to define `Type` as a union type:
```
Type = PrimitiveType | Array | TaggedUnion
```

File: jdwp/common.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    """Types class."""
+
+    STRING = str
+    INT = int
+    BYTE = float
+    BOOLEAN = bool
+    DICT = dict
```
Comment (michalgr) [inline jdwp/common.py:13]: This one seems odd

File: jdwp/common.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    INTERFACE_ID = "interfaceID"
+    CLASS_OBJECT_ID = "classObjectID"
+    TAGGED_OBJECT_ID = "taggedObjectID"
+    THREAD_ID = "threadID"
+    THREAD_GROUP_ID = "threadGroupID"
+    OBJECT_ID = "objectID"
+    LOCATION = "location"
+
+
+class Field:
```
Comment (michalgr) [inline jdwp/common.py:28]: It might be more convenient to use `NamedTuple`s or dataclasses here and later. They are immutable, less verbose to declare and have whole bunch of useful methods out of the box:

```
class Field(typing.NamedTuple):
    name: str
    type: Type
    description: str
```

**Review by michalgr — COMMENTED** [review]
Great start !

I have a suggestion to organize code a bit differently. I would:
- move Type, Field, Struct, Command and Command set to `projects/jdwp/defs/schema.py`
- move all remaining enums to `projects/jdwp/defs/constants.py`
- move definitions of command sets to `projects/jdwp/defs/commandsets/...`

File: jdwp/common.py (hunk 2 above)
Comment (Daquiver1) [reply jdwp/common.py:13]: Yeah, I've made modifications to the code. However, I have reservations about how we will accurately retrieve the specific datatype. Currently, I've implemented it as a string that holds the datatype's value. I'm uncertain if that works.

File: projects/jdwp/defs/schema.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    fields: List[Field]
+
+
+@dataclass(frozen=True)
+class Command:
+    """Command class."""
+
+    name: str
+    id: int
+    out: Struct
```
Comment (michalgr) [inline projects/jdwp/defs/schema.py:67]: There are commands that don't have any `out` data.

```
out: Optional[Struct]
```

File: projects/jdwp/defs/schema.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    reply: Struct
```
Comment (michalgr) [inline projects/jdwp/defs/schema.py:68]: Similarly here
```
reply: Optional[Struct]
```

File: projects/jdwp/defs/schema.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    error: Struct
```
Comment (michalgr) [inline projects/jdwp/defs/schema.py:69]: Let's either remove this or turn this into a sequence/set of allowed errors. For example:
```
error: collections.abc.Set[ErrorKind]
```

File: projects/jdwp/defs/schema.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+
+@dataclass(frozen=True)
+class CommandSet:
+    """Command set class."""
+
+    name: str
+    id: int
+    commands: List[Command]
```
Comment (michalgr) [inline projects/jdwp/defs/schema.py:78]: It's better to type this as `collections.abc.Sequence[Command]`, it's more generic and excludes mutability.

File: projects/jdwp/defs/schema.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    name: str
+    type = Types
+    description: str
+
+
+@dataclass(frozen=True)
+class Struct:
+    """Struct class."""
+
+    fields: List[Field]
```
Comment (michalgr) [inline projects/jdwp/defs/schema.py:58]:
```
fields: collections.abc.Sequence[Field]
```
(see below for explanations)

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (michalgr) [inline .vscode/settings.json:1]: I'd prefer to enable `black` the same way as `pyre`. If we enable `black` then we also need to make sure out jdwp github actions run pyre check and fail if code is not formatted properly.

I would make this kind of change in separate PR.

File: projects/jdwp/constants/errors.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+"""Errors enum class."""
+from enum import Enum
+
+
+class ErrorConstants(Enum):
```
Comment (michalgr) [inline projects/jdwp/constants/errors.py:5]: Perhaps `ErrorKind` or `ErrorType` would be a better name (singular form, `Sequence[ErrorKind]` is more informative than `Sequence[ErrorConstants]` which suggests a sequence in which each element consists of potentially multiple error constants)

File: projects/jdwp/constants/invoke_options.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+"""Invoke options enum class."""
+from enum import IntEnum
+
+
+class InvokeOptions(IntEnum):
```
Comment (michalgr) [inline projects/jdwp/constants/invoke_options.py:5]: Should this be a `enum.Flag` instead of `IntEnum` ?

**Review by michalgr — CHANGES_REQUESTED** [review]
Not all commands are correctly described. In particular, some of them should have array fields and this PR does not add support for Arrays. Let's extract a part of this PR that can be landed sooner and will unblock @wekesa360.

Could you please do the following:
- address feedback in `schema.py` and `constants/*.py`
- move all definitions of constants to `defs/constants.py`
- for now remove
  - `defs/commands/thread_reference.py` and `defs/command_sets/thread_reference.py`
  - all commands from`defs/commands/reference_type.py` but the `Signature` command
- move definition of `Signature` from `defs/commands/reference_type.py` to `defs/command_sets/reference_type.py`
- make `ReferenceType` command set incomplete, consisting of only one command (`Signature`) for now
- rebase on top of main and make sure code builds with buck and checks with pyre.

**Comment by michalgr** [comment]
Could you also start files with this copyright ?
https://github.com/facebookexperimental/ExtendedAndroidTools/blob/main/projects/jdwp/main.py#L1

**Comment by michalgr** [comment]
Almost perfect!

Pyre does not type check new files because they are not part of any buck target. Could you add
`projects/jdwp/defs/BUCK` file with the following content ?
```
# Copyright (c) Meta Platforms, Inc. and affiliates.

python_library(
    name = "defs",
    srcs = glob(["**/*.py"]),
    visibility = ["PUBLIC"],
)
```

This will make all the new files visible to buck and pyre.

File: projects/jdwp/BUCK [hunk 12]
```diff
@@ -8,7 +8,7 @@ python_binary(
 
 
 python_library(
-    name = "lib",
-    srcs = [],
+    name = "defs",
```
Comment (michalgr) [inline projects/jdwp/BUCK:11]: This is not `projects/jdwp/defs/BUCK`, but that is fine. In this case please keep the original name ("lib") and we should be good to go.
